openvdb/tools/FastSweeping.h

```diff
@@ -0,0 +1,1432 @@
+///          is a scalar fog volume (i.e. higher values are inside), and Ext is
+///          a field (currently limited to a scalar) that is extended off the
+///          iso-surface. All these functions are implemented by the methods of
+///          the class dubbed FastSweeping.
+///
+/// @todo    1) Sort uint32 offsets instead of array of Coord
+///          2) Concurrent bi-directional sweeping
+///          3) Rebuild narrow-band level set
+///          4) Allow for the grid types of the sdf and ext to be different
+///
+/// @note    Solves the (simplified) Eikonal Eq: @f$|\nabla \phi|^2 = 1@f$ and
+///          performs velocity extension,  @f$\nable f\nabla \phi = 0@f$, both
+///          by means of the fast sweeping algorithm detailed in:
+///          "A Fast Sweeping Method For Eikonal Equations"
+///          by H. Zhao, Mathematics of Computation, Vol 74(230), pp 603-627, 2004
+///
+/// @details The algorithm used below for parallel fast sweeping was first publised in:
+///          "New Algorithm for Sparse and Parallel Fast Sweeping: Efficient
+///          Computation of Sparse Distance Fields" by K. Museth, ACM SIGGRAPH Talk,
+///          2017, http://www.museth.org/Ken/Publications_files/Museth_SIG17.pdf
+
+#ifndef OPENVDB_TOOLS_FASTSWEEPING_HAS_BEEN_INCLUDED
+#define OPENVDB_TOOLS_FASTSWEEPING_HAS_BEEN_INCLUDED
+
+//#define BENCHMARK_FAST_SWEEPING
+
+#include <type_traits>// for static_assert
+#include <cmath>
+#include <limits>
+#include <thread>
+
+#include <tbb/blocked_range.h>
+#include <tbb/parallel_for.h>
+#include <tbb/parallel_sort.h>
+#include <tbb/parallel_invoke.h>
+#include <tbb/enumerable_thread_specific.h>
+
+#include <openvdb/math/Math.h> // for Abs() and isExactlyEqual()
+#include <openvdb/math/Stencils.h> // for GradStencil
+#include <openvdb/util/PagedArray.h>
+#include <openvdb/tree/LeafManager.h>
+#include "LevelSetUtil.h"
+#include "Morphology.h"
+
+#include "Statistics.h"
+#ifdef BENCHMARK_FAST_SWEEPING
+#include <openvdb/util/CpuTimer.h>
+#endif
+
+#include <openvdb/math/Stats.h>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace tools {
+
+/// @brief Converts a scalar fog volume into a signed distance function. Active input voxels
+///        with scalar values above the given isoValue will have NEGATIVE distance
+///        values on output, i.e. they are assumed to be INSIDE the iso-surface.
+///
+/// @return A shared pointer to a signed-distance field defined on the active values
+///         of the input fog volume.
+///
+/// @param fogGrid  Scalar (floating-point) volume from which an
+///                 iso-surface can be defined.
+///
+/// @param isoValue A value which defines a smooth iso-surface that
+///                 intersects active voxels in @a fogGrid.
+///
+/// @param nIter    Number of iterations of the fast sweeping algorithm.
+///                 each performing 2^3 = 8 sweeps.
```

Comment:
I think I mentioned in the meeting that as future work it would be cool to have a tolerance for convergence...

Reply:
I agree, but the problem is to come up with a metric or tolerance. The norm of the gradient would seem like a good starting point but the first-order nature of this fast sweeping method implies that there's a limit to the accuracy and the existence of singularities (critical points) might complicated things as well. In other words, it's far from trivial.

Reply:
Yeah, the gradient is unlikely to converge.  I was thinking maybe just absolute change could be used.  When the largest voxel change update less than a certain percentage of voxel diameter?   Are the sweeps monotonic?  If so, this should at least always stop (even if it does so by stalling out far from the final values)

Reply:
this sill be added later